lms/djangoapps/instructor/enrollment.py

```diff
@@ -219,26 +219,41 @@ def reset_student_attempts(course_id, student, module_state_key, delete_module=F
         submissions.SubmissionError: unexpected error occurred while resetting the score in the submissions API.
 
     """
+    user_id = anonymous_id_for_user(student, course_id)
+    requesting_user_id = anonymous_id_for_user(requesting_user, course_id)
+    submission_cleared = False
     try:
         # A block may have children. Clear state on children first.
         block = modulestore().get_item(module_state_key)
         if block.has_children:
             for child in block.children:
                 try:
-                    reset_student_attempts(course_id, student, child, delete_module=delete_module)
+                    reset_student_attempts(course_id, student, child, requesting_user, delete_module=delete_module)
                 except StudentModule.DoesNotExist:
                     # If a particular child doesn't have any state, no big deal, as long as the parent does.
                     pass
+        if delete_module:
+            # Some blocks (openassessment) use StudentModule data as a key for internal submission data.
+            # Inform these blocks of the reset and allow them to handle their data.
+            clear_student_state = getattr(block, "clear_student_state", None)
+            if callable(clear_student_state):
+                clear_student_state(
+                    user_id=user_id,
+                    course_id=unicode(course_id),
+                    item_id=unicode(module_state_key),
+                    requesting_user_id=requesting_user_id
+                )
+                submission_cleared = True
     except ItemNotFoundError:
         log.warning("Could not find %s in modulestore when attempting to reset attempts.", module_state_key)
 
-    # Reset the student's score in the submissions API
-    # Currently this is used only by open assessment (ORA 2)
-    # We need to do this *before* retrieving the `StudentModule` model,
-    # because it's possible for a score to exist even if no student module exists.
-    if delete_module:
+    # Reset the student's score in the submissions API, if xblock.clear_student_state has not done so already.
+    # TODO: Remove this once we've finalized and communicated how xblocks should handle clear_student_state
+    # and made sure that other xblocks relying on the submission api understand this is going away.
+    # We need to do this before retrieving the `StudentModule` model, because a score may exist with no student module.
```

Comment:
Can you move the "TODO: ..." below the line starting "We need to..."?  Also, can you update the TODO as follows?:
```
# TODO: Should the LMS know about sub_api and call this reset, or should it generically call it on all of its xblock services as well?  See JIRA ARCH-26.
```